gltf-utils/src/lib.rs

```diff
@@ -59,147 +42,185 @@ pub trait PrimitiveIterators<'a> {
-    ) -> Option<WeightsF32<'s>>;
+    fn weights<'s, S>(&'a self, set: u32, source: &'s S) -> Option<Weights<'s>>
+        where S: Source;
 }
 
 impl<'a> PrimitiveIterators<'a> for gltf::Primitive<'a> {
-    fn positions<'s, S: Source>(&self, source: &'s S) -> Option<Positions<'s>> {
+    fn positions<'s, S: Source>(&'a self, source: &'s S) -> Option<Positions<'s>> {
         self.get(&gltf::Semantic::Positions)
-            .map(|accessor| Positions(AccessorIter::new(accessor, source)))
+            .map(|accessor| AccessorIter::new(accessor, source))
     }
 
-    fn normals<'s, S: Source>(&self, source: &'s S) -> Option<Normals<'s>> {
+    fn normals<'s, S: Source>(&'a self, source: &'s S) -> Option<Normals<'s>> {
         self.get(&gltf::Semantic::Normals)
-            .map(|accessor| Normals(AccessorIter::new(accessor, source)))
+            .map(|accessor| AccessorIter::new(accessor, source))
     }
 
-    fn tangents<'s, S: Source>(&self, source: &'s S) -> Option<Tangents<'s>> {
+    fn tangents<'s, S: Source>(&'a self, source: &'s S) -> Option<Tangents<'s>> {
         self.get(&gltf::Semantic::Tangents)
-            .map(|accessor| Tangents(AccessorIter::new(accessor, source)))
+            .map(|accessor| AccessorIter::new(accessor, source))
     }
 
-    fn tex_coords_f32<'s, S: Source>(&self, set: u32, source: &'s S) -> Option<TexCoordsF32<'s>> {
-        self.get(&gltf::Semantic::TexCoords(set))
-            .map(|accessor| TexCoordsF32(TexCoords::new(accessor, source)))
-    }
+    fn colors<'s, S>(&'a self, set: u32, source: &'s S) -> Option<Colors<'s>>
+        where S: Source
+    {
+        use DataType::{U8, U16, F32};
+        use Dimensions::{Vec3, Vec4};
 
-    fn colors_rgba_f32<'s, S: Source>(
-        &self,
-        set: u32,
-        default_alpha: f32,
-        source: &'s S,
-    ) -> Option<ColorsRgbaF32<'s>> {
         self.get(&gltf::Semantic::Colors(set))
-            .map(|accessor| {
-                ColorsRgbaF32 {
-                    iter: Colors::new(accessor, source),
-                    default_alpha,
-                }
+            .map(|accessor| match (accessor.data_type(), accessor.dimensions()) {
+                (U8, Vec3)  => Colors::RgbU8(AccessorIter::new(accessor, source)),
+                (U16, Vec3) => Colors::RgbU16(AccessorIter::new(accessor, source)),
+                (F32, Vec3) => Colors::RgbF32(AccessorIter::new(accessor, source)),
+                (U8, Vec4)  => Colors::RgbaU8(AccessorIter::new(accessor, source)),
+                (U16, Vec4) => Colors::RgbaU16(AccessorIter::new(accessor, source)),
+                (F32, Vec4) => Colors::RgbaF32(AccessorIter::new(accessor, source)),
+                _ => unreachable!(),
             })
     }
 
-    fn indices_u32<'s, S: Source>(&self, source: &'s S) -> Option<IndicesU32<'s>> {
-        self.indices().map(|accessor| IndicesU32(Indices::new(accessor, source)))
+    fn indices<'s, S>(&'a self, source: &'s S) -> Option<Indices<'s>>
```

Comment:
Don't worry about the name clash. One can use uniform function call syntax if necessary.

Reply:
Well, yes, but it's not very nice.  Shadowing Primitive::indices with PrimitiveIterators::indices requires that you to `use` it from a scope higher than Primitive is used at.